### PR TITLE
Implement multi-range apply

### PR DESCRIPTION
**Pull request by ihnorton**

### CODE REVIEW

File: tiledb/cloud/cloudarray.py [hunk 1]
```diff
@@ -144,3 +163,32 @@ def apply(self, func, subarray, attrs=None, layout=None):
...
         return cloudpickle.loads(res)
+
+    def _calculate_ranges(self, subarray):
+        ranges = list()
+        if isinstance(subarray, MultiIndexRanges):
+            for dim_idx, dim_rng in enumerate(subarray.ranges):
+                dim = self.schema.domain.dim(dim_idx)
+                for idx, rng in enumerate(dim_rng):
+                    start = build_dimension_coordinate(dim.dtype, rng[0])
+                    end = build_dimension_coordinate(dim.dtype, rng[1] + 1)
```
Comment (Shelnutt2) [inline tiledb/cloud/cloudarray.py]: Should we be adding a +1 here?

Comment (ihnorton) [reply tiledb/cloud/cloudarray.py]: nope, thanks -- left over from testing on prod when it was endpoint-exclusive

Comment (ihnorton) [reply tiledb/cloud/cloudarray.py]: adding some tests for apply now

**Comment by ihnorton** [comment]
@stavrospapadopoulos please see the simplified syntax in the new tests, here:

https://github.com/TileDB-Inc/TileDB-Cloud-Py/blob/44eadefcf6d181ebf414a3900699a51056dc494c/tests/test_basic.py#L31-L41

**Pull request merged**